lib/pubsub/subscription.js

```diff
@@ -278,50 +287,98 @@ Subscription.prototype.delete = function(callback) {
 
 /**
  * Pull messages from the subscribed topic. If messages were found, your
- * callback is executed with the message object.
+ * callback is executed with an array of message objects.
  *
  * Note that messages are pulled automatically once you register your first
  * event listener to the subscription, thus the call to `pull` is handled for
  * you. If you don't want to start pulling, simply don't register a
  * `subscription.on('message', function() {})` event handler.
  *
+ * @todo Should not be racing with other pull.
+ * @todo Fix API to return a list of messages.
+ *
  * @param {object=} options - Configuration object.
- * @param {boolean=} options.returnImmediately - If set, the system will respond
+ * @param {boolean} options.returnImmediately - If set, the system will respond
  *     immediately. Otherwise, wait until new messages are available. Returns if
  *     timeout is reached.
+ * @param {number} options.maxCount - Limit the amount of messages pulled.
  * @param {function} callback - The callback function.
  *
  * @example
- * subscription.pull(function(err, message) {
- *   // message.id = ID used to acknowledge its receival.
- *   // message.data = Contents of the message.
+ * //-
+ * // Pull all available messages.
+ * //-
+ * subscription.pull(function(err, messages) {
+ *   // messages = [
+ *   //   {
+ *   //     ackId: '', // ID used to acknowledge its receival.
+ *   //     id: '',    // Unique message ID.
+ *   //     data: ''   // Contents of the message.
+ *   //   },
+ *   //   // ...
+ *   // ]
+ * });
+ *
+ * //-
+ * // Pull a single message.
+ * //-
+ * var opts = {
+ *   maxCount: 1
+ * };
+ *
+ * subscription.pull(opts, function(err, message) {
+ *   // message = {
+ *   //   ackId: '', // ID used to acknowledge its receival.
+ *   //   id: '',    // Unique message ID.
+ *   //   data: ''   // Contents of the message.
+ *   // }
  * });
  */
 Subscription.prototype.pull = function(options, callback) {
   var that = this;
-  // TODO(jbd): Should not be racing with other pull.
+
   if (!callback) {
     callback = options;
     options = {};
   }
+
+  if (!util.is(options.maxCount, 'number')) {
+    options.maxCount = 99999;
+  }
+
+  var single = options.maxCount === 1;
+  var apiEndpoint = single ? 'subscriptions/pull' : 'subscriptions/pullBatch';
```

Comment:
The only reason we still have pull, is not to break old code. We introduced pullBatch and publishBatch without changing the version.

So please just use only pullBatch and publishBatch.

Thanks!